launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java

```diff
@@ -200,6 +200,7 @@ void addOptionString(List<String> cmd, String options) {
 
     addToClassPath(cp, getenv("HADOOP_CONF_DIR"));
     addToClassPath(cp, getenv("YARN_CONF_DIR"));
+    addToClassPath(cp, getEffectiveConfig().get("spark.yarn.conf.dir"));
```

Comment:
I'm wondering how do we update the classpath to change to another hadoop confs with InProcessLauncher? Seems the classpath here is not changeable after JVM is launched.

Reply:
On another note, is this meant to extend to other resource-managers? As Kubernetes assumes only the ENV `HADOOP_CONF_DIR`, but if such a change is desirable this would cause a slight re-work of the current Hadoop Conf Files mounting logic.

Reply:
@jerryshao My understanding is that this method is not used by the InProcessLauncher. So instead, the caller of InProcessLauncher has to make sure that the conf files are available to hadoop's Configuration class in the YarnClusterApplication. For example, by adding the config files to the calling thread's context class loader

Reply:
Saisai's question about the classpath configuration is actually the most complicated part of this feature. I haven't fully thought about how they would play out, but I really don't think it's as simple as appending this new config to the classpath.

e.g. what is the expectation if you run "spark-shell" with this option? Do you end up using the config from the env variable or from the config? If you have both, and you reference a file in `--files` that is on an HDFS namespace declared in the `hdfs-site.xml` from the config, what will happen? (Answer: it will be ignored, since that is being masked by the `hdfs-site.xml` from the env variable.)

Reply:
Yes, it is quite tricky. My expectation is that it would behave the same way as if you pointed `HADOOP_CONF_DIR` and `YARN_CONF_DIR` to different directories that both contain `hdfs-site.xml`. Files in `HADOOP_CONF_DIR` would take precedence (as far as I know, nothing prevents this from happening). So with this new config, the order of preference would be `HADOOP_CONF_DIR`, `YARN_CONF_DIR`, then `spark.yarn.conf.dir`. 

Perhaps this could be clarified in the docs,  but let me know what you think about it, I'm happy to implement other resolutions.

Reply:
@vanzin Did you have time to think about how this config should work?